pkg/controller/planexecution/plan_execution_engine_test.go

```diff
@@ -0,0 +1,153 @@
+		operatorVersion:     "ov-1.0",
+		operatorName:        "operator",
+		resourcesOwner:      getJob("pod2", "default"),
+		operatorVersionName: "ovname",
+	}
+	tests := []struct {
+		name           string
+		activePlan     *activePlan
+		metadata       *executionMetadata
+		expectedStatus *v1alpha1.PlanExecutionStatus
+	}{
+		{"plan already finished", &activePlan{
+			Name: "test",
+			State: &v1alpha1.PlanExecutionStatus{
+				State: v1alpha1.PhaseStateComplete,
+			},
+		}, defaultMetadata, &v1alpha1.PlanExecutionStatus{
+			State: v1alpha1.PhaseStateComplete,
+		}},
+		{"plan with one step to be executed, still in progress", &activePlan{
+			Name: "test",
+			State: &v1alpha1.PlanExecutionStatus{
+				State:    v1alpha1.PhaseStatePending,
+				Name:     "test",
+				Strategy: "serial",
+				Phases:   []v1alpha1.PhaseStatus{{Strategy: "serial", Name: "phase", State: v1alpha1.PhaseStatePending, Steps: []v1alpha1.StepStatus{{State: v1alpha1.PhaseStatePending, Name: "step"}}}},
+			},
+			Spec: &v1alpha1.Plan{
+				Strategy: "serial",
+				Phases: []v1alpha1.Phase{
+					{Name: "phase", Strategy: "serial", Steps: []v1alpha1.Step{{Name: "step", Tasks: []string{"task"}}}},
+				},
+			},
+			Tasks:     map[string]v1alpha1.TaskSpec{"task": {Resources: []string{"job"}}},
+			Templates: map[string]string{"job": getResourceAsString(getJob("job1", "default"))},
+		}, defaultMetadata, &v1alpha1.PlanExecutionStatus{
+			State:    v1alpha1.PhaseStatePending,
+			Name:     "test",
+			Strategy: "serial",
+			Phases:   []v1alpha1.PhaseStatus{{Strategy: "serial", Name: "phase", State: v1alpha1.PhaseStateInProgress, Steps: []v1alpha1.StepStatus{{State: v1alpha1.PhaseStateInProgress, Name: "step"}}}},
+		}},
+		// this plan deploys pod, that is marked as healthy immediately because we cannot evaluate health
+		{"plan with one step, immediately healthy -> completed", &activePlan{
+			Name: "test",
+			State: &v1alpha1.PlanExecutionStatus{
+				State:    v1alpha1.PhaseStatePending,
+				Name:     "test",
+				Strategy: "serial",
+				Phases:   []v1alpha1.PhaseStatus{{Strategy: "serial", Name: "phase", State: v1alpha1.PhaseStatePending, Steps: []v1alpha1.StepStatus{{State: v1alpha1.PhaseStatePending, Name: "step"}}}},
+			},
+			Spec: &v1alpha1.Plan{
+				Strategy: "serial",
+				Phases: []v1alpha1.Phase{
+					{Name: "phase", Strategy: "serial", Steps: []v1alpha1.Step{{Name: "step", Tasks: []string{"task"}}}},
+				},
+			},
+			Tasks:     map[string]v1alpha1.TaskSpec{"task": {Resources: []string{"pod"}}},
+			Templates: map[string]string{"pod": getResourceAsString(getPod("pod1", "default"))},
+		}, defaultMetadata, &v1alpha1.PlanExecutionStatus{
+			State:    v1alpha1.PhaseStateComplete,
+			Name:     "test",
+			Strategy: "serial",
+			Phases:   []v1alpha1.PhaseStatus{{Strategy: "serial", Name: "phase", State: v1alpha1.PhaseStateComplete, Steps: []v1alpha1.StepStatus{{State: v1alpha1.PhaseStateComplete, Name: "step"}}}},
+		}},
+	}
```

Comment:
I cannot really test the non-retrying on fatal because that is what the controller is doing (it's either returning the error to the loop or not). I'll write a test to make sure when this method is called on a state that is errored, we'll retry.

Reply:
that fatal error handling is happening here https://github.com/kudobuilder/kudo/pull/759/files#diff-c2e52ee1aa92f961c976702674f98591R338

From the execution standpoint it's safe to retry... we just should not do it to not get into infinite loop